### PR TITLE
Add Spring Security and user authentification

### PR DESCRIPTION
**Pull request by Max-Leopold**
Add user authentification through spring security to the application.

You can't view queries or mentions without being logged in. Queries are now bound to user accounts. You can only retrieve queries and their mentions if you wrote the query.

When logged in Spring saves the Credentials of the user in the browser. When a call is made  the `Spring Security Context` is sent to the endpoint inside the Http Request.

Example:

`Session Attrs = {SPRING_SECURITY_CONTEXT=org.springframework.security.core.context.SecurityContextImpl@fd6e26c: Authentication: org.springframework.security.authentication.UsernamePasswordAuthenticationToken@fd6e26c: Principal: org.springframework.security.core.userdetails.User@f02988d6: Username: username; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Not granted any authorities; Credentials: [PROTECTED]; Authenticated: true; Details: org.springframework.security.web.authentication.WebAuthenticationDetails@957e: RemoteIpAddress: 127.0.0.1; SessionId: null; Not granted any authorities}`

### CODE REVIEW

File: API/src/main/java/de/brandwatch/minianalytics/api/security/service/CustomUserDetailsService.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Autowired
+    public CustomUserDetailsService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByUsername(username);
+        if(user == null) throw new UsernameNotFoundException(username);
```
Comment (lfzCarlosC) [inline API/src/main/java/de/brandwatch/minianalytics/api/security/service/CustomUserDetailsService.java]: formatting issue. put throw in another line

File: API/src/main/java/de/brandwatch/minianalytics/api/service/MentionService.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        User user = userRepository.findByUsername(username);
+
+        if(!queryRepository.findById(Long.valueOf(queryID)).get().getUser().equals(user)){
+            throw new Exception("Query " + queryID + " doesnt belong to user " + username);
+        }
+
+        StringBuilder sb = new StringBuilder();
+        sb.append("[");
+
+        if(!endDate.equals("") && !startDate.equals("")){
```
Comment (lfzCarlosC) [inline API/src/main/java/de/brandwatch/minianalytics/api/service/MentionService.java]: Not sure if you need to check these two dates are empty or null. For the empty case, better be :
`!"".equals(endDate) && !"".equals(startDate)`

Comment (Max-Leopold) [reply API/src/main/java/de/brandwatch/minianalytics/api/service/MentionService.java]: Isn't `!endDate.equals("")` and `!"".equals(endDate)` the exact same? Just asking for clarification.

File: API/src/main/java/de/brandwatch/minianalytics/api/security/service/CustomUserDetailsService.java (hunk 1 above)
Comment (Max-Leopold) [reply API/src/main/java/de/brandwatch/minianalytics/api/security/service/CustomUserDetailsService.java]: Changed.

**Pull request merged**